docs/RFCS/20222202_historical_hot_ranges.md

```diff
@@ -0,0 +1,224 @@
+
+To provide meaningful resolution, we can set a sampling rate of 4 times per hour, and store samples for a period of 14 days.
+
+For each range, here is the information captured:
+| Value      | Bytes |  description |
+| ---------- |------ | ---      |
+| node_id    | 2    |   | 
+| store_id   | 2    |  |
+| qps        | 4    |  |
+| start_key  |  <= 100  | locate the start of the hot range |
+| end_key    | <= 100  | locate the end of the hot range |
+| timestamp  | 4 | seconds since epoch |
+
+
+We can calculate the total storage requirements of this data, less the cost of the indexes and a unique identifier, considering all dimensions:
+```
+<= 212 bytes per range * 1000 ranges per sample * 4 samples per hour * 24 hours * 14 days = <= 285 Megabytes
+```
+
+This value could be much less depending on the distribution of `start_key` and `end_key` character lengths.
+
+
+It's important to point out that the client will never need the majority of these bytes to be sent over the wire in order to render the visualization. Instead, the client will request samples in chunks, with subsequent API calls made to request the specific details of a range only as needed. This is discussed in more detail later on.
+
+
+___
+### SQL Schema
+
+Capping the storage requirement to 285 Megabytes, we can leverage CRDB's internal tables to store HHR.
+The table below provides the proposed schema for `crdb_internal.historical_hot_ranges`. To efficiently query ranges for a given sample, the `timestamp` column is given a secondary index. To efficiently query a range by its start key, `start_key` is given a secondary index. The need for this is discussed in the "Querying" section later on.
+
+| column name | type | primary index | secondary index
+| ---------- |------ | --- | --- |
+| uuid       | UUID | yes  | |
+| node_id    | INT2  |   | |
+| store_id    | INT2  |   | |
+| qps    | FLOAT4  |   | |
+| start_key    | STRING(100)  |   | yes |
+| end_key    | STRING(100)  |   | |
+| timestamp    | INT4  |   | yes |
+
+
+___
+### Metric Generation 
+
+According to the sample rate, every node in the cluster will collect metrics for a subset of leaseholder replicas that live on its stores. Then, every node will persist these metrics to `crdb_internal.historical_hot_ranges`. The stores are responsible for determining the subset and aggregation of leaseholder replicas that are recorded. Every store operates with a budget to guarantee that no more than 1000 ranges are recorded cumulatively by all nodes. At the time of sampling, existing samples older than 2 weeks will be deleted.
```

Comment:
As an FYI, @andreimatei is leading the discussions around an approach where nodes will report their metrics to special servers that will aggregate these metrics and persist to storage. In any case, we need to make sure this works well in multi-region, where the cost of a write might be expensive if the leaseholder for that row is across WAN.

Reply:
One potential benefit of pre-aggregation is that we would perhaps be able to consolidate colder ranges into larger spans, further reducing the storage requirements for such ranges (users probably don't care about our range boundaries, although they might care about table or database boundaries).

Reply:
Thinking out-loud here, another potential benefit is if we can create very coarse-grained spans that can be presented initially by the dashboard. This will allow showing something quickly. Then, as the user zooms in, or maybe just gradually with time, we can improve the presented granularity.